### DISCUSSION TITLE
Поле с маской ввода

### DISCUSSION THREAD

**Discussion by OAndrey1985** [totumonline/totum-ru-issues-and-discussions#484] [Как это сделать?] [answered]
Ищу в документации, но не могу привязаться к поисковому запросу, как правильно организовать поле ввода по маске, тот же номер телефона, например "+1" жестко потом пробел и  "(" вводим код ну и далее по схеме пользователь набирает цифры а записывается +1 (812) 321-22-48 понимаю как то через код, пните в нужное место.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Лайфхак по regexp — спросить у GPT

Сделай мне regexp пропускающий такой формат телефона:

+X (XXX) XXX-XX-XX

Да, конечно! Вот пример регулярного выражения (RegExp) для формата телефона +X (XXX) XXX-XX-XX:

```
^\+\d \(\d{3}\) \d{3}-\d{2}-\d{2}$
```

Пояснение:

- ^: начало строки.
- \+: символ "+" должен быть экранирован, так как он имеет специальное значение в регулярных выражениях.
- \d: любая цифра.
- \( и \): экранированные скобки для обозначения открывающей и закрывающей скобок соответственно.
- \d{3}: три цифры подряд.
- -: символ тире.
- $: конец строки.

Таким образом, это регулярное выражение будет соответствовать формату телефона +X (XXX) XXX-XX-XX, где X - любая цифра.

**Reply by ttm-alexeypolunin** [reply]
Парметр маски в поле называется Фомат валидизации ввода: https://ru.docs.totum.online/fields#fields-settings-regexp

**Reply by ttm-alexeypolunin** [reply]
Вот еще пример из реального проекта по приведению номеров при вводе в форматированный вид.

регексп задан как `^[0-9+() -]+$` — только цифры и +-() и пробел

Потом кодом действия на изменение просим GPT сделать нам красиво и делаем set обратно в поле:

```
= : if(condition: $#nfv != ""; condition: $@ask = ""; then: $while)

	while: while(action: $var; action: $set)
		var: procVar(name: "ask"; value: true)

set: set(table: $#ntn; field: $#nf = $answer[answer]; log: "ChatGPT")

	answer: serviceAskOpenai(system: "Отформатируй номер в формате +Х XXX XXX-XX-XX верни только номер в json в ключе answer" = $json; system: "Если первый символ 8 то заменяй его на +7";  user: $#nfv; answer: "json")
	
		json: rowCreate(field: "answer" = "")
```

https://www.youtube.com/watch?v=yJ8SD91Y9aU

**Reply by OAndrey1985** [reply]
Гениальное приложение, спасибо, вроде понятно

**Reply by ttm-alexeypolunin** [reply]
Забыл пояснить (тк это редко используется) про это `$@ask = ""`.

Дело в том, что при `set` тоже будет изменение и произойдет повторный запрос к Сервису, чего не нужно.

Поэтому при первом запросе мы создаем Процессную переменную, которая существует только в экземпляре `fpm` выполняющего этот код. Вот так:

```
while: while(action: $var; action: $set)
		var: procVar(name: "ask"; value: true)
```

Ф. `while` здесь используется для выполнения двух последовательных действий — первым мы создаем переменную и записываем в нее `true`.

Когда придет `set` и код действия запустится снова и проверив переменную можно определить это первый запуск или повторный.

**Comment by OAndrey1985** [comment]
> nswer" = $json; sys

А можно regexp задать в коде, при условии 1 такой то, при условии 2 такое то условие, вдруг пригодится. Синтаксис предусматривает такую возможность?

**Comment by OAndrey1985** [comment]
> ^\+\d \(\d{3}\) \d{3}-\d{2}-\d{2}$

Попробовал по маске сделать, в настроках поля указываешь и как бы если не знать маску +1 и т.д. то он автоматом не подставляет символы, только надо плейсхолдер сделать в виде подсказки а символы автоматически при добавлении цифр не добавляются и тире тоже, чтобы например набирать 7905445454 а он подставлял  +  и - в нужные места
